src/goto-symex/symex_dereference_state.cpp

```diff
@@ -13,6 +13,21 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include <util/symbol_table.h>
 
+/// Get or create a failed symbol for the given pointer-typed expression. These
+/// are used as placeholders when dereferencing expressions that are illegal to
+/// dereference, such as null pointers. The \ref add_failed_symbols pass must
+/// have been run for this to do anything useful;  it annotates a pointer-typed
+/// symbol `p` with an `ID_C_failed_symbol` comment, which we then clone on
+/// demand due to L1 renaming.
+///
+/// For example, if \p expr is `p` and it has an `ID_C_failed_symbol` `p$object`
+/// (the naming convention used by `add_failed_symbols`), and the latest L1
+/// renaming of `p` is `p!2@4`, then we will create `p$object!2@4` if it doesn't
+/// already exist.
+///
+/// \param expr: expression to get or create a failed symbol for
+/// \param [out] symbol: failed symbol result
+/// \return true if we populated \p symbol
 bool symex_dereference_statet::get_or_create_failed_symbol(
```

Comment:
Not sure whether in this PR or in a separate one, but could we please change the return value to `optionalt<symbolt>`? The "Boolean return value + out-parameter" combination is just too error-prone and nowadays we have better options.

Reply:
Pushed an extra commit to just use `symbolt *`, since that's what the existing code was expecting